FrozenIndexIT.testRetryPointInTime

May be this is just one off failure, and should be ignored. Please feel free to close

**Build scan**:
https://gradle-enterprise.elastic.co/s/3tlvmqrujbggq

**Repro line**:
./gradlew ':x-pack:plugin:frozen-indices:internalClusterTest' --tests "org.elasticsearch.index.engine.FrozenIndexIT.testRetryPointInTime" -Dtests.seed=DCA39E4A873000C6 -Dtests.security.manager=true -Dtests.locale=ar-LB -Dtests.timezone=Europe/Skopje -Druntime.java=11	

**Reproduces locally?**:
no

**Applicable branches**:
was failed on PR for master branch

**Failure history**:
<!--
Link to build stats and possible indication of when this started failing and how often it fails
<https://build-stats.elastic.co/app/kibana>
-->
**Failure excerpt**:
```bash

java.lang.AssertionError: Unexpected ShardFailures: [shard [[UYO248ViTFGMo9H2xCimwQ][test][0]], reason [org.elasticsearch.transport.RemoteTransportException: [node_s0][127.0.0.1:34427][indices:data/read/search[phase/fetch/id]] |  
-- | --
  | Caused by: org.elasticsearch.search.SearchContextMissingException: No search context found for id [1] |  
  | at org.elasticsearch.search.SearchService.findReaderContext(SearchService.java:623) |  
  | at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:586) |  
  | at org.elasticsearch.action.search.SearchTransportService.lambda$registerRequestHandler$12(SearchTransportService.java:363) |  
  | at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:72) |  
  | at org.elasticsearch.transport.TransportService.sendLocalRequest(TransportService.java:805) |  
  | at org.elasticsearch.transport.TransportService$3.sendRequest(TransportService.java:125) |  
  | at org.elasticsearch.transport.TransportService.sendRequestInternal(TransportService.java:749) |  
  | at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:663) |  
  | at org.elasticsearch.transport.TransportService.sendChildRequest(TransportService.java:714) |  
  | at org.elasticsearch.transport.TransportService.sendChildRequest(TransportService.java:706) |  
  | at org.elasticsearch.action.search.SearchTransportService.sendExecuteFetch(SearchTransportService.java:182) |  
  | at org.elasticsearch.action.search.SearchTransportService.sendExecuteFetch(SearchTransportService.java:172) |  
  | at org.elasticsearch.action.search.FetchSearchPhase.executeFetch(FetchSearchPhase.java:176) |  
  | at org.elasticsearch.action.search.FetchSearchPhase.innerRun(FetchSearchPhase.java:156) |  
  | at org.elasticsearch.action.search.FetchSearchPhase$1.doRun(FetchSearchPhase.java:95) |  
  | at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) |  
  | at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:44) |  
  | at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:739) |  
  | at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) |  
  | at java.base/java.lang.Thread.run(Thread.java:834) |  
  | ], cause [org.elasticsearch.search.SearchContextMissingException: No search context found for id [1] |  
  | at org.elasticsearch.search.SearchService.findReaderContext(SearchService.java:623) |  
  | at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:586) |  
  | at org.elasticsearch.action.search.SearchTransportService.lambda$registerRequestHandler$12(SearchTransportService.java:363) |  
  | at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:72) |  
  | at org.elasticsearch.transport.TransportService.sendLocalRequest(TransportService.java:805) |  
  | at org.elasticsearch.transport.TransportService$3.sendRequest(TransportService.java:125) |  
  | at org.elasticsearch.transport.TransportService.sendRequestInternal(TransportService.java:749) |  
  | at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:663) |  
  | at org.elasticsearch.transport.TransportService.sendChildRequest(TransportService.java:714) |  
  | at org.elasticsearch.transport.TransportService.sendChildRequest(TransportService.java:706) |  
  | at org.elasticsearch.action.search.SearchTransportService.sendExecuteFetch(SearchTransportService.java:182) |  
  | at org.elasticsearch.action.search.SearchTransportService.sendExecuteFetch(SearchTransportService.java:172) |  
  | at org.elasticsearch.action.search.FetchSearchPhase.executeFetch(FetchSearchPhase.java:176) |  
  | at org.elasticsearch.action.search.FetchSearchPhase.innerRun(FetchSearchPhase.java:156) |  
  | at org.elasticsearch.action.search.FetchSearchPhase$1.doRun(FetchSearchPhase.java:95) |  
  | at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) |  
  | at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:44) |  
  | at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:739) |  
  | at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) |  
  | at java.base/java.lang.Thread.run(Thread.java:834) |  
  | ]] |  
  | Expected: <0> |  
  | but: was <1>


```

--- a/server/src/main/java/org/elasticsearch/search/SearchService.java
+++ b/server/src/main/java/org/elasticsearch/search/SearchService.java
@@ -648,18 +648,18 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
                     searcherSupplier.close();
                     throw e;
                 }
-                return createAndPutReaderContext(request, indexService, shard, searcherSupplier, false);
+                return createAndPutReaderContext(request, indexService, shard, searcherSupplier, false, defaultKeepAlive);
             }
         } else {
             final IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
             final IndexShard shard = indexService.getShard(request.shardId().id());
             final Engine.SearcherSupplier searcherSupplier = shard.acquireSearcherSupplier();
-            return createAndPutReaderContext(request, indexService, shard, searcherSupplier, keepStatesInContext);
+            return createAndPutReaderContext(request, indexService, shard, searcherSupplier, keepStatesInContext, getKeepAlive(request));
         }
     }
 
     final ReaderContext createAndPutReaderContext(ShardSearchRequest request, IndexService indexService, IndexShard shard,
-                                                  Engine.SearcherSupplier reader, boolean keepStatesInContext) {
+                                                  Engine.SearcherSupplier reader, boolean keepStatesInContext, long keepAliveInMillis) {
         ReaderContext readerContext = null;
         Releasable decreaseScrollContexts = null;
         try {
@@ -672,16 +672,15 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
                             + MAX_OPEN_SCROLL_CONTEXT.getKey() + "] setting.");
                 }
             }
-            final long keepAlive = getKeepAlive(request);
             final ShardSearchContextId id = new ShardSearchContextId(sessionId, idGenerator.incrementAndGet());
             if (keepStatesInContext || request.scroll() != null) {
-                readerContext = new LegacyReaderContext(id, indexService, shard, reader, request, keepAlive);
+                readerContext = new LegacyReaderContext(id, indexService, shard, reader, request, keepAliveInMillis);
                 if (request.scroll() != null) {
                     readerContext.addOnClose(decreaseScrollContexts);
                     decreaseScrollContexts = null;
                 }
             } else {
-                readerContext = new ReaderContext(id, indexService, shard, reader, keepAlive, request.keepAlive() == null);
+                readerContext = new ReaderContext(id, indexService, shard, reader, keepAliveInMillis, true);
             }
             reader = null;
             final ReaderContext finalReaderContext = readerContext;

--- a/server/src/main/java/org/elasticsearch/search/SearchService.java
+++ b/server/src/main/java/org/elasticsearch/search/SearchService.java
@@ -651,10 +651,11 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
                 return createAndPutReaderContext(request, indexService, shard, searcherSupplier, false, defaultKeepAlive);
             }
         } else {
+            final long keepAliveInMillis = getKeepAlive(request);
             final IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
             final IndexShard shard = indexService.getShard(request.shardId().id());
             final Engine.SearcherSupplier searcherSupplier = shard.acquireSearcherSupplier();
-            return createAndPutReaderContext(request, indexService, shard, searcherSupplier, keepStatesInContext, getKeepAlive(request));
+            return createAndPutReaderContext(request, indexService, shard, searcherSupplier, keepStatesInContext, keepAliveInMillis);
         }
     }
 

--- a/server/src/test/java/org/elasticsearch/search/SearchServiceTests.java
+++ b/server/src/test/java/org/elasticsearch/search/SearchServiceTests.java
@@ -583,7 +583,7 @@ public class SearchServiceTests extends ESSingleNodeTestCase {
         final ShardScrollRequestTest request = new ShardScrollRequestTest(indexShard.shardId());
         ElasticsearchException ex = expectThrows(ElasticsearchException.class,
             () -> service.createAndPutReaderContext(
-                request, indexService, indexShard, indexShard.acquireSearcherSupplier(), randomBoolean()));
+                request, indexService, indexShard, indexShard.acquireSearcherSupplier(), randomBoolean(), request.keepAlive().millis()));
         assertEquals(
             "Trying to create too many scroll contexts. Must be less than or equal to: [" +
                 SearchService.MAX_OPEN_SCROLL_CONTEXT.get(Settings.EMPTY) + "]. " +
@@ -611,8 +611,9 @@ public class SearchServiceTests extends ESSingleNodeTestCase {
                     for (; ; ) {
                         final Engine.SearcherSupplier reader = indexShard.acquireSearcherSupplier();
                         try {
-                            searchService.createAndPutReaderContext(
-                                new ShardScrollRequestTest(indexShard.shardId()), indexService, indexShard, reader, true);
+                            final ShardScrollRequestTest request = new ShardScrollRequestTest(indexShard.shardId());
+                            searchService.createAndPutReaderContext(request, indexService, indexShard, reader, true,
+                                request.keepAlive().millis());
                         } catch (ElasticsearchException e) {
                             assertThat(e.getMessage(), equalTo(
                                 "Trying to create too many scroll contexts. Must be less than or equal to: " +
@@ -1119,7 +1120,8 @@ public class SearchServiceTests extends ESSingleNodeTestCase {
                         OriginalIndices.NONE, new SearchRequest().allowPartialSearchResults(true),
                         indexShard.shardId(), 0, 1, new AliasFilter(null, Strings.EMPTY_ARRAY), 1.0f, -1, null);
                     final ReaderContext context = searchService.createAndPutReaderContext(request, indexService, indexShard,
-                        indexShard.acquireSearcherSupplier(), randomBoolean());
+                        indexShard.acquireSearcherSupplier(), randomBoolean(),
+                        SearchService.KEEPALIVE_INTERVAL_SETTING.get(Settings.EMPTY).millis());
                     assertThat(context.id().getId(), equalTo((long) (i + 1)));
                     contextIds.add(context.id());
                 }

--- a/server/src/test/java/org/elasticsearch/search/SearchServiceTests.java
+++ b/server/src/test/java/org/elasticsearch/search/SearchServiceTests.java
@@ -582,8 +582,8 @@ public class SearchServiceTests extends ESSingleNodeTestCase {
 
         final ShardScrollRequestTest request = new ShardScrollRequestTest(indexShard.shardId());
         ElasticsearchException ex = expectThrows(ElasticsearchException.class,
-            () -> service.createAndPutReaderContext(
-                request, indexService, indexShard, indexShard.acquireSearcherSupplier(), randomBoolean(), request.keepAlive().millis()));
+            () -> service.createAndPutReaderContext(request, indexService, indexShard, indexShard.acquireSearcherSupplier(),
+                randomBoolean(), SearchService.KEEPALIVE_INTERVAL_SETTING.get(Settings.EMPTY).millis()));
         assertEquals(
             "Trying to create too many scroll contexts. Must be less than or equal to: [" +
                 SearchService.MAX_OPEN_SCROLL_CONTEXT.get(Settings.EMPTY) + "]. " +
@@ -613,7 +613,7 @@ public class SearchServiceTests extends ESSingleNodeTestCase {
                         try {
                             final ShardScrollRequestTest request = new ShardScrollRequestTest(indexShard.shardId());
                             searchService.createAndPutReaderContext(request, indexService, indexShard, reader, true,
-                                request.keepAlive().millis());
+                                SearchService.KEEPALIVE_INTERVAL_SETTING.get(Settings.EMPTY).millis());
                         } catch (ElasticsearchException e) {
                             assertThat(e.getMessage(), equalTo(
                                 "Trying to create too many scroll contexts. Must be less than or equal to: " +

--- a/x-pack/plugin/frozen-indices/src/internalClusterTest/java/org/elasticsearch/index/engine/FrozenIndexIT.java
+++ b/x-pack/plugin/frozen-indices/src/internalClusterTest/java/org/elasticsearch/index/engine/FrozenIndexIT.java
@@ -184,7 +184,6 @@ public class FrozenIndexIT extends ESIntegTestCase {
         }
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/67307")
     public void testRetryPointInTime() throws Exception {
         internalCluster().ensureAtLeastNumDataNodes(1);
         final List<String> dataNodes =

